// react-sweet-state/lib/cjs/store/create-state.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _middlewares = _interopRequireDefault(require("../middlewares"));

var _devtools = _interopRequireDefault(require("../enhancers/devtools"));

var _defaults = _interopRequireDefault(require("../defaults"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function createStoreState(key, initialState) {
  var _listeners2 = [];
  var currentState = initialState;
  var storeState = {
    key: key,
    getState: function getState() {
      return currentState;
    },
    setState: function setState(nextState) {
      currentState = nextState;

      for (var _i = 0, _listeners = _listeners2; _i < _listeners.length; _i++) {
        var listener = _listeners[_i];
        listener(currentState);
      }
    },
    resetState: function resetState() {
      storeState.setState(initialState);
    },
    subscribe: function subscribe(listener) {
      _listeners2 = _listeners2.concat(listener);
      return function unsubscribe() {
        _listeners2 = _listeners2.filter(function (fn) {
          return fn !== listener;
        });
      };
    },
    listeners: function listeners() {
      return _listeners2;
    }
  };
  storeState.mutator = (0, _middlewares["default"])(storeState, _defaults["default"].middlewares);
  return storeState;
}

var _default = (0, _devtools["default"])(createStoreState);

exports["default"] = _default;
